tools/rust_analyzer/rust_project.rs

```diff
@@ -0,0 +1,304 @@
+
+use anyhow::anyhow;
+use serde::Serialize;
+
+use crate::aquery::CrateSpec;
+
+/// A `rust-project.json` workspace representation. See
+/// [rust-analyzer documentation][rd] for a thorough description of this interface.
+/// [rd]: https://rust-analyzer.github.io/manual.html#non-cargo-based-projects
+#[derive(Serialize)]
+pub struct RustProject {
+    /// Path to the directory with *source code* of
+    /// sysroot crates.
+    sysroot_src: Option<String>,
+
+    /// The set of crates comprising the current
+    /// project. Must include all transitive
+    /// dependencies as well as sysroot crate (libstd,
+    /// libcore and such).
+    crates: Vec<Crate>,
+}
+
+/// A `rust-project.json` crate representation. See
+/// [rust-analyzer documentation][rd] for a thorough description of this interface.
+/// [rd]: https://rust-analyzer.github.io/manual.html#non-cargo-based-projects
+#[derive(Serialize)]
+pub struct Crate {
+    /// A name used in the package's project declaration
+    #[serde(skip_serializing_if = "Option::is_none")]
+    display_name: Option<String>,
+
+    /// Path to the root module of the crate.
+    root_module: String,
+
+    /// Edition of the crate.
+    edition: String,
+
+    /// Dependencies
+    deps: Vec<Dependency>,
+
+    /// Should this crate be treated as a member of current "workspace".
+    #[serde(skip_serializing_if = "Option::is_none")]
+    is_workspace_member: Option<bool>,
+
+    /// Optionally specify the (super)set of `.rs` files comprising this crate.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    source: Option<Source>,
+
+    /// The set of cfgs activated for a given crate, like
+    /// `["unix", "feature=\"foo\"", "feature=\"bar\""]`.
+    cfg: Vec<String>,
+
+    /// Target triple for this Crate.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    target: Option<String>,
+
+    /// Environment variables, used for the `env!` macro
+    #[serde(skip_serializing_if = "Option::is_none")]
+    env: Option<HashMap<String, String>>,
+
+    /// Whether the crate is a proc-macro crate.
+    is_proc_macro: bool,
+
+    /// For proc-macro crates, path to compiled proc-macro (.so file).
+    #[serde(skip_serializing_if = "Option::is_none")]
+    proc_macro_dylib_path: Option<String>,
+}
+
+#[derive(Serialize)]
+pub struct Source {
+    include_dirs: Vec<String>,
+    exclude_dirs: Vec<String>,
+}
+
+#[derive(Serialize)]
+pub struct Dependency {
+    /// Index of a crate in the `crates` array.
+    #[serde(rename = "crate")]
+    crate_index: usize,
+
+    /// The display name of the crate.
+    name: String,
+}
+
+pub fn generate_rust_project(
+    sysroot_src: &str,
+    crates: &[CrateSpec],
+) -> anyhow::Result<RustProject> {
+    let mut project = RustProject {
+        sysroot_src: Some(sysroot_src.into()),
+        crates: Vec::new(),
+    };
+
+    let mut unmerged_crates: Vec<&CrateSpec> = crates.iter().collect();
+    let mut skipped_crates: Vec<&CrateSpec> = Vec::new();
```

Comment:
It's getting too late here for my brain to think, but it seems this algorithm is going to great lengths to process a crate after its deps and do it all in a single loop. But the only reason for that seems to be to get the index of the dep. And that sounds like unnecessary requirement. Maybe the index could be precomputed in one linear pass, and then the crates could be generated in another linear pass?

Also I'm not sure I see where the merging happens? As noted below, could you add a test that shows the scenario of rust_library and rust_test using `crate` attribute (therefore they have the same id I believe)?

Reply:
Yes, that's the goal. It iteratively processes any crate with zero dependencies that have not yet been populated into the list. I had to do some things that felt a little goofy to satisfy the borrow checker, so there may be ways to simplify it. The complexity is worst case n^2, but only occurs when the dependency tree is entirely linear and reverse ordered. In most cases the complexity is closer to n*d where d is the maximal depth of the dependency tree, which is generally a fairly small constant, and n collapses very quickly after 1-3 passes as most library crates have a fairly shallow dependency tree.

I am not sure that you can do this in two linear passes, because I think that rust-analyzer may assume that dependencies always appear prior to use, but I am not sure of this. I _believe_ that's why I chose this route, but I am not certain of that -- there were a few bugs during development, and I may have been dealing with a different bug that turned out to be that some dependencies were not getting populated at all. So, the two linear passes approach may actually be fine, it just needs to be tested to ensure out-of-order dependencies work with rust-analyzer.

Reply:
Discussed with Andre offline, I'm fine with leaving this investigation/change to a follow-up PR.